### PR TITLE
Make IGrainFactory injectable

### PR DESCRIPTION
**Pull request by xiazen**
Make IGrainFactory injectable. 

In the future we want to make more internal system types injectable through DI container, for application layer to use and for making orleans' DI better. This is the first step.

### CODE REVIEW

File: src/OrleansRuntime/Silo/Silo.cs [hunk 1]
```diff
@@ -234,7 +235,19 @@ internal Silo(string name, SiloType siloType, ClusterConfiguration config, ILoca
             AppDomain.CurrentDomain.UnhandledException +=
                 (obj, ev) => DomainUnobservedExceptionHandler(obj, (Exception)ev.ExceptionObject);
 
-            grainFactory = new GrainFactory();
+            // we are registering GrainFactory as IGrainFactory in StartupBuilder.cs, 
+            // and we are not providing the ability for application developer to replace IGrainFactory
```
Comment (ReubenBond) [inline src/OrleansRuntime/Silo/Silo.cs]: This comment doesn't match the code

Comment (xiazen) [inline src/OrleansRuntime/Silo/Silo.cs]: thanks. fixed it :)

File: src/OrleansRuntime/Silo/Silo.cs [hunk 2]
```diff
@@ -234,7 +235,16 @@ internal Silo(string name, SiloType siloType, ClusterConfiguration config, ILoca
...
                 (obj, ev) => DomainUnobservedExceptionHandler(obj, (Exception)ev.ExceptionObject);
 
-            grainFactory = new GrainFactory();
+            try
+            {
+                grainFactory = Services.GetRequiredService<GrainFactory>();
+            }
+            catch (InvalidOperationException exc)
+            {
+                logger.Error(ErrorCode.SiloStartError, "Exception during Silo.Start, GrainFactory not regiested in Dependency Injection container", exc);
```
Comment (chbomtempo) [inline src/OrleansRuntime/Silo/Silo.cs]: There is a typo on the exception message. "regiested"

Comment (xiazen) [inline src/OrleansRuntime/Silo/Silo.cs]: thanks ! fixed it !

**Comment by xiazen** [comment]
someone take a look and merge it ?

**Pull request merged**